homeassistant/components/ondilo_ico/sensor.py

```diff
@@ -0,0 +1,185 @@
+
+        return pool
+
+    async def async_update_data():
+        """Fetch data from API endpoint.
+
+        This is the place to pre-process the data to lookup tables
+        so entities can quickly look up their data.
+        """
+        try:
+            pools = await hass.async_add_executor_job(api.get_pools)
+
+            return await asyncio.gather(
+                *[
+                    hass.async_add_executor_job(get_all_pool_data, pool)
+                    for pool in pools
+                ]
+            )
+
+        except OndiloError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        # Name of the data. For logging purposes.
+        name="sensor",
+        update_method=async_update_data,
+        # Polling interval. Will only be polled if there are subscribers.
+        update_interval=SCAN_INTERVAL,
+    )
+
+    # Fetch initial data so we have data when entities subscribe
+    await coordinator.async_refresh()
+
+    entities = []
+    for poolidx, pool in enumerate(coordinator.data):
+        for sensor_idx, sensor in enumerate(pool["sensors"]):
+            if sensor["data_type"] in SENSOR_TYPES:
+                entities.append(OndiloICO(coordinator, poolidx, sensor_idx))
+
+    async_add_entities(entities)
+
+
+class OndiloICO(CoordinatorEntity):
+    """Representation of a Sensor."""
+
+    def __init__(
+        self, coordinator: DataUpdateCoordinator, poolidx: int, sensor_idx: int
+    ):
+        """Initialize sensor entity with data from coordinator."""
+        super().__init__(coordinator)
+
+        self._poolid = self.coordinator.data[poolidx]["id"]
+
+        pooldata = self._pooldata()
+        self._data_type = pooldata["sensors"][sensor_idx]["data_type"]
+        self._unique_id = f"{pooldata['ICO']['serial_number']}-{self._data_type}"
+        self._device_name = pooldata["name"]
+        self._name = f"{self._device_name} {SENSOR_TYPES[self._data_type][0]}"
+        self._device_class = SENSOR_TYPES[self._data_type][3]
+        self._icon = SENSOR_TYPES[self._data_type][2]
+        self._unit = SENSOR_TYPES[self._data_type][1]
+
+    def _pooldata(self):
+        """Get pool data dict."""
+        return next(
+            (pool for pool in self.coordinator.data if pool["id"] == self._poolid),
+            None,
+        )
+
+    def _devdata(self):
+        """Get device data dict."""
+        return next(
+            (
+                data_type
+                for data_type in self._pooldata()["sensors"]
+                if data_type["data_type"] == self._data_type
+            ),
+            None,
+        )
+
+    @property
+    def name(self):
+        """Name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Last value of the sensor."""
+        _LOGGER.debug(
```

Comment:
We don't do side effects, like logging, in entity state properties. Please remove the logging.

Reply:
Done in PR https://github.com/home-assistant/core/pull/44837